task.go

```diff
@@ -34,28 +36,77 @@ var (
 		http.StatusTemporaryRedirect: true, // 307
 		http.StatusPermanentRedirect: true, // 308
 	}
+
+	errReferenceLoop = errors.New("detected cyclic reference")
 )
 
-type packagerTask struct {
+type packagerTaskRunner struct {
 	*Packager
+
+	period *exchange.ValidPeriod
+	errs   *multierror.MultiError
+	active map[string]bool // Keyed by URLs.
+}
+
+func newTaskRunner(p *Packager, vp *exchange.ValidPeriod) *packagerTaskRunner {
+	return &packagerTaskRunner{
+		p,
+		vp,
+		&multierror.MultiError{},
+		make(map[string]bool),
+	}
+}
+
+func (runner *packagerTaskRunner) err() error {
+	return runner.errs.Err()
+}
+
+func (runner *packagerTaskRunner) run(parent *packagerTask, req *http.Request, r *resource.Resource) {
+	url := r.RequestURL.String()
+	var err error
+
+	if runner.active[url] {
+		err = errReferenceLoop
+	} else {
+		log.Printf("processing %v ...", url)
+		runner.active[url] = true
+		err = (&packagerTask{runner, parent, req, r}).run()
+		delete(runner.active, url)
+	}
+
+	if err != nil {
+		err = fmt.Errorf("error with processing %s: %v", url, err)
```

Comment:
I believe `%w` is more favorable in `fmt.Errorf` as of Go 1.13 (ref: https://blog.golang.org/go1.13-errors)

Reply:
I don't think `%w` works in Go 1.11 & 1.12 though...